### PR TITLE
Contact us form captcha

### PR DESCRIPTION
**Pull request by rBangay**
## What does this change?
Adds a re-captcha gate to the contact us form.

## Images
Screenshot 2020-11-05 at 10 07 46

Screenshot 2020-11-05 at 10 03 29

### CODE REVIEW

**Review by ripecosta — APPROVED** [review]
👍

**Review by ripecosta — CHANGES_REQUESTED** [review]
The reCaptcha key needs to be submitted with the rest of the form so it can be validated server-side. Otherwise we will be vulnerable to spamming if someone sends messages directly to our endpoint.

**Comment by rBangay** [comment]
> The reCaptcha key needs to be submitted with the rest of the form so it can be validated server-side. Otherwise we will be vulnerable to spamming if someone sends messages directly to our endpoint.

The form should now pass along the captcha token `captchaToken` along with the rest of the form data

File: app/server/contactUsApi.ts [hunk 1]
```diff
@@ -65,7 +65,9 @@ const validateContactUsFormBody = (body: any): boolean =>
   body.email &&
   isEmail(body.email) &&
   body.subject &&
-  body.message;
+  body.message &&
+  body &&
```
Comment (ripecosta) [inline app/server/contactUsApi.ts:69]: Line 61 already checks `body` so this one is not needed.
```suggestion
```

File: app/server/contactUsApi.ts [hunk 2]
```diff
@@ -65,7 +65,9 @@ const validateContactUsFormBody = (body: any): boolean =>
...
+  body.captchaToken;
```
Comment (ripecosta) [inline app/server/contactUsApi.ts:70]: The token itself also need to be validated here.

**Review by ripecosta — CHANGES_REQUESTED** [review]

File: app/server/contactUsApi.ts [hunk 3]
```diff
@@ -50,18 +51,39 @@ export const contactUsFormHandler = async (req: Request, res: Response) => {
...
 
-const parseAndValidate = (body: any): ContactUsReq | undefined => {
+const parseAndValidate = async (
+  body: any
+): Promise<ContactUsReq | undefined> => {
   try {
     const bodyAsJson = body ? JSON.parse(body) : "{}";
 
-    return validateContactUsFormBody(bodyAsJson)
+    const passedCaptchaValidation = await validateCaptchaToken(
```
Comment (ripecosta) [inline app/server/contactUsApi.ts:60]: At this point in the execution it still hasn't been confirmed that `bodyAsJson.captchaToken` exists. `validateContactUsFormBody` is the validate step of this function, so it makes more sense to validate the captcha inside that function.

File: app/server/contactUsApi.ts [hunk 4]
```diff
@@ -50,18 +51,39 @@ export const contactUsFormHandler = async (req: Request, res: Response) => {
...
+    return validateContactUsFormBody(bodyAsJson) && passedCaptchaValidation
       ? buildContactUsReqBody(bodyAsJson)
       : undefined;
   } catch (error) {
     return undefined;
   }
 };
 
+const validateCaptchaToken = async (token: string) => {
+  const recaptchaSecret = (await recaptchaConfigPromise)?.secretKey;
```
Comment (ripecosta) [inline app/server/contactUsApi.ts:72]: Can we fail the check before making the validation fetch if `recaptchaSecret` is undefined?

File: app/server/recaptchaConfig.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { s3ConfigPromise } from "./awsIntegration";
+
+interface ReCaptchaKeys {
+  publicKey: string;
+  secretKey: string;
+}
+
+export const recaptchaConfigPromise = s3ConfigPromise<ReCaptchaKeys>()(
```
Comment (ripecosta) [inline app/server/recaptchaConfig.ts:8]: Should we log an error in Sentry if the promise comes back with 'undefined'?

**Review by ripecosta — CHANGES_REQUESTED** [review]

**Review by ripecosta — APPROVED** [review]
👍

**Pull request merged**